File: erty-backend/src/models/Transaction.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const Transaction = sequelize.define('Transaction', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true
  },
  userId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  accountId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'accounts',
      key: 'id'
    }
  },
  categoryId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'categories',
      key: 'id'
    }
  },
  type: {
    type: DataTypes.ENUM('expense', 'income', 'transfer'),
    allowNull: false
  },
  amount: {
    type: DataTypes.DECIMAL(15, 2),
    allowNull: false,
    validate: {
      min: 0
    }
  },
  currency: {
    type: DataTypes.STRING,
    defaultValue: 'AOA'
  },
  description: {
    type: DataTypes.TEXT
  },
  date: {
    type: DataTypes.DATE,
    allowNull: false,
    defaultValue: DataTypes.NOW
  },
  paymentMethod: {
    type: DataTypes.ENUM(
      'cash',
      'debit_card',
      'credit_card',
      'bank_transfer',
      'pix',
      'mobile_money',
      'other'
    ),
    defaultValue: 'cash'
  },
  status: {
    type: DataTypes.ENUM('pending', 'completed', 'cancelled'),
    defaultValue: 'completed'
  },
  isRecurring: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  recurringId: {
    type: DataTypes.UUID,
    allowNull: true,
    references: {
      model: 'recurring_transactions',
      key: 'id'
    }
  },
  toAccountId: {
    type: DataTypes.UUID,
    allowNull: true, // Usado em transferências
    references: {
      model: 'accounts',
      key: 'id'
    }
  },
  attachments: {
    type: DataTypes.JSONB,
    defaultValue: []
  },
  location: {
    type: DataTypes.JSONB,
    defaultValue: null // { latitude, longitude, address }
  },
  tags: {
    type: DataTypes.ARRAY(DataTypes.STRING),
    defaultValue: []
  },
  notes: {
    type: DataTypes.TEXT
  },
  isSynced: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  syncedAt: {
    type: DataTypes.DATE
  },
  localId: {
    type: DataTypes.STRING, // ID local do dispositivo (para sincronização offline)
    unique: true
  },
  aiSuggested: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  aiConfidence: {
    type: DataTypes.DECIMAL(3, 2),
    validate: {
      min: 0,
      max: 1
    }
  },
  metadata: {
    type: DataTypes.JSONB,
    defaultValue: {}
  }
});

module.exports = Transaction;

File: erty-mobile/lib/domain/entities/transaction.dart
import 'package:equatable/equatable.dart';

enum TransactionType { expense, income, transfer }

class Transaction extends Equatable {
  final String id;
  final String userId;
  final String accountId;
  final String categoryId;
  final TransactionType type;
  final double amount;
  final String currency;
  final String? description;
  final DateTime date;
  final String paymentMethod;
  final String status;
  final List<String> tags;
  final String? notes;
  final bool isSynced;
  final String? localId;

  const Transaction({
    required this.id,
    required this.userId,
    required this.accountId,
    required this.categoryId,
    required this.type,
    required this.amount,
    required this.currency,
    this.description,
    required this.date,
    required this.paymentMethod,
    required this.status,
    this.tags = const [],
    this.notes,
    this.isSynced = false,
    this.localId,
  });

  @override
  List<Object?> get props => [
        id,
        userId,
        accountId,
        categoryId,
        type,
        amount,
        currency,
        description,
        date,
        paymentMethod,
        status,
        tags,
        notes,
        isSynced,
        localId,
      ];

  Transaction copyWith({
    String? id,
    String? userId,
    String? accountId,
    String? categoryId,
    TransactionType? type,
    double? amount,
    String? currency,
    String? description,
    DateTime? date,
    String? paymentMethod,
    String? status,
    List<String>? tags,
    String? notes,
    bool? isSynced,
    String? localId,
  }) {
    return Transaction(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      accountId: accountId ?? this.accountId,
      categoryId: categoryId ?? this.categoryId,
      type: type ?? this.type,
      amount: amount ?? this.amount,
      currency: currency ?? this.currency,
      description: description ?? this.description,
      date: date ?? this.date,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      status: status ?? this.status,
      tags: tags ?? this.tags,
      notes: notes ?? this.notes,
      isSynced: isSynced ?? this.isSynced,
      localId: localId ?? this.localId,
    );
  }
}
